doc: Add docs for the old assumptions (#20090)
sympy/sympy: doc/src/tutorial/assumptions.rst

    True
    >>> expr2 = expr.expand()
    >>> expr2
    x**2 - 4*x + 5
    >>> print(expr2.is_positive)
    None

Ideally that last example would give ``True`` rather than ``None`` because the
expression is always positive for any real value of ``x`` (and ``x`` has been
assumed real). The assumptions system is intended to be efficient though: it
is expected many more complex queries will not be fully resolved. This is
because assumptions queries are primarily used internally by SymPy as part of
low-level calculations. Making the system more comprehensive would slow SymPy
down.

Note that in fuzzy logic giving an indeterminate result ``None`` is never a
contradiction. If it is possible to infer a definite ``True`` or ``False``
result when resolving a query then that is better than returning ``None``.
However a result of ``None`` is not a *bug*. Any code that uses the
assumptions system needs to be prepared to handle all three cases for any
query and should not presume that a definite answer will always be given.

The assumptions system is not just for symbols or for complex expressions. It
can also be used for plain SymPy integers and other objects. The assumptions

> predicates are available on any instance of ``Basic`` which is the superclass
> for most classes of SymPy objects. A plain Python ``int`` is not a ``Basic``
> instance and will can not be used to query assumptions predicates. We can
> "sympify" regular Python objects to become SymPy objects with ``sympify`` or
> ``S`` and then the assumptions system can be used:

Reviewer: Might be sensible to link to the docs here
```suggestion
predicates are available on any instance of :class:`Basic` which is the superclass
for most classes of SymPy objects. A plain Python :class:`int` is not a ``Basic``
instance and will can not be used to query assumptions predicates. We can
"sympify" regular Python objects to become SymPy objects with "func:`sympify` or
:data:`S` and then the assumptions system can be used:
```

    >>> from sympy import S
    >>> x = 2
    >>> x.is_positive
    Traceback (most recent call last):
    ...
    AttributeError: 'int' object has no attribute 'is_positive'
    >>> x = S(2)
    >>> type(x)
    <class 'sympy.core.numbers.Integer'>
    >>> x.is_positive
    True

Gotcha: symbols with different assumptions
==========================================

In SymPy it is possible to declare two symbols with different names and they
will implicitly be considered equal under *structural equality*:

    >>> x1 = Symbol('x')
    >>> x2 = Symbol('x')
    >>> x1
    x
    >>> x2
    x
    >>> x1 == x2
    True

However if the symbols have different assumptions then they will be considered
to represent distinct symbols:

    >>> x1 = Symbol('x', positive=True)
    >>> x2 = Symbol('x', )
    >>> x1
    x
    >>> x2
    x
    >>> x1 == x2
    False

> One way to simplify an expression is to use the ``posify`` function which will

Reviewer: ```suggestion
One way to simplify an expression is to use the :func:`posify` function which will
```

replace all symbols in an expression with symbols that have the assumption
``positive=True`` (unless that contradicts any existing assumptions for the
symbol):

    >>> from sympy import posify, exp
    >>> x = Symbol('x')
    >>> expr = exp(sqrt(x**2))
    >>> expr
    exp(sqrt(x**2))
    >>> posify(expr)
    (exp(_x), {_x: x})
    >>> expr2, rep = posify(expr)
    >>> expr2
    exp(_x)

The ``posify`` function returns the expression with all symbols replaced
(which might lead to simplifications) and also a dict which maps the new

> symbols to the old that can be used with ``.subs``. This is useful because

Reviewer: ```suggestion
symbols to the old that can be used with :meth:`~Basic.subs`. This is useful because
```

otherwise the new expression with the new symbols having the ``positive=True``
assumption will not compare equal to the old:

    >>> expr2
    exp(_x)
    >>> expr2 == exp(x)
    False
    >>> expr2.subs(rep)
    exp(x)
    >>> expr2.subs(rep) == exp(x)
    True

Applying assumptions to string inputs
=====================================

> We have seen how to set assumptions when calling the ``Symbol`` or
> ``symbols`` functions explicitly. A natural question to ask is: in what other

Reviewer: ```suggestion
We have seen how to set assumptions when calling the :class:`Symbol` or
:func:`symbols` functions explicitly. A natural question to ask is: in what other
```

situations can we assign assumptions to an object?

It is common for users to use strings as input to SymPy functions (although
the general feeling among SymPy developers is that this should be discouraged)
e.g.:

    >>> from sympy import solve
    >>> solve('x**2 - 1')
    [-1, 1]

When creating symbols explicitly it would be possible to assign assumptions
that would affect the behaviour of ``solve``:

    >>> x = Symbol('x', positive=True)
    >>> solve(x**2 - 1)
    [1]

When using string input SymPy will create the expression and create all of the
symbolc implicitly so the question arises how can the assumptions be
specified? The answer is that rather than depending on implicit string

> cponversion it is better to use the ``parse_expr`` function explicitly and

Reviewer: ```suggestion
conversion it is better to use the :func:`parse_expr` function explicitly and
```

then it is possible to provide assumptions for the symbols e.g.:

    >>> from sympy import parse_expr
    >>> parse_expr('x**2 - 1')
    x**2 - 1
    >>> eq = parse_expr('x**2 - 1', {'x':Symbol('x', positive=True)})
    >>> solve(eq)
    [1]

> .. note:: The solve function is unusual as a high level API in that it

Reviewer: ```suggestion
.. note:: The :func:`solve` function is unusual as a high level API in that it
```

          actually checks the assumptions on any input symbols (the unknowns)
          and uses that to tailor its output. The assumptions system otherwise
          affects low-level evaluation but is not necessarily handled
          explicitly by high-level APIs.

Predicates
==========

There are many different predicates that can be assumed for a symbol or can be
queried for an expression. It is possible to combine multiple predicates when
creating a symbol. Predicates are logically combined using *and* so if a
symbol is declared with ``positive=True`` and also with ``integer=True`` then
it is both positive *and* integer:

    >>> x = Symbol('x', positive=True, integer=True)
    >>> x.is_positive
    True
    >>> x.is_integer
    True

The full set of known predicates for a symbol can be accessed using the

> ``assumptions0`` attribute:

Reviewer: ```suggestion
:attr:`Basic.assumptions0` attribute:
```

number (in this context even SymPy's standard infinities are considered
"ordinary").

The only predicate that is applied by default for a symbol is ``commutative``.
We can also declare a symbol to be *noncommutative* e.g.:

    >>> x, y = symbols('x, y', commutative=False)
    >>> z = Symbol('z')  # defaults to commutative=True
    >>> x*y + y*x
    x*y + y*x
    >>> x*z + z*x
    2*z*x

Note here that since ``x`` and ``y`` are both noncommutative ``x`` and ``y``
do not commute so ``x*y != y*x``. On the other hand since ``z`` is commutative
``x`` and ``z`` commute and ``x*z == z*x`` even though ``z`` is
noncommutative.

The interpretation of what a vanilla symbol represents is unclear but the
interpretation of an expression with ``commutative=False`` is entirely
obscure. Such an expression is necessarily not a complex number or an
extended real or any of the standard infinities (even ``zoo`` is commutative).
We are left with very little that we can say about what such an expression
*does* represent.

Symbolic Boolean vs fuzzy bool
==============================

Assumptions queries give fuzzy bool ``True``, ``False`` or ``None`` results.

> These are low-level Python objects rather than SymPy's symbolic ``Boolean``

Reviewer: ```suggestion
These are low-level Python objects rather than SymPy's symbolic :class:`Boolean`
```

Author: Turns out `Boolean` isn't actually documented anywhere

expressions. As an example an inequality in SymPy is a ``Boolean`` and can
represent indeterminate results symbolically:

    >>> xpos = Symbol('xpos', positive=True)
    >>> xneg = Symbol('xneg', negative=True)
    >>> x = Symbol('x')
    >>> print(xpos.is_positive)
    True
    >>> print(xneg.is_positive)
    False
    >>> print(x.is_positive)
    None
    >>> xpos > 0
    True
    >>> xneg > 0
    False
    >>> x > 0
    x > 0

The last example shows what happens when an inequality is indeterminate: we

> get an instance of ``StrictGreaterThan`` which represents the inequality as a

Reviewer: ```suggestion
get an instance of :class:`StrictGreaterThan`` which represents the inequality as a
```

is returned as show for ``x > 0`` above.

It is not obvious that queries like ``xpos > 0`` return ``S.true`` rather than
``True`` because both objects display in the same way but we can check this
using the Python ``is`` operator:

    >>> from sympy import S
    >>> xpos.is_positive is True
    True
    >>> xpos.is_positive is S.true
    False
    >>> (xpos > 0) is True
    False
    >>> (xpos > 0) is S.true
    True

There is no general symbolic analogue of ``None`` in SymPy. In the cases where
a low-level assumptions query gives ``None`` the symbolic query will result in
an unevaluated symbolic ``Boolean`` (e.g, ``x > 0``).  We can use a symbolic
``Boolean`` as part of a symbolic expression:

    >>> from sympy import Piecewise
    >>> p = Piecewise((1, x > 0), (2, True))
    >>> p
    Piecewise((1, x > 0), (2, True))
    >>> p.subs(x, 3)
    1

Here ``p`` represents an expression that will be equal to ``1`` if ``x > 0``
or otherwise it will be equal to ``2``. The unevaluated ``Boolean`` inequality
``x > 0`` represents the condition for deciding the value of the expression
symbolically. When we substitute a value for ``x`` the inequality will resolve

> to ``S.true`` and then the ``Piecewise`` can evaluate to ``1`` or ``2``.

Reviewer: ```suggestion
to ``S.true`` and then the :class:`Piecewise` can evaluate to ``1`` or ``2``.
```

The same will not work when using a fuzzy-bool instead of a symbolic
``Boolean``:

    >>> p2 = Piecewise((1, x.is_positive), (2, True))
    Traceback (most recent call last):
    ...
    TypeError: Second argument must be a Boolean, not `NoneType`

The ``Piecewise`` can not use ``None`` as the condition because unlike the
inequality ``x > 0`` it gives no information. With the inequality it is
possible to decide in future if the condition might ``True`` or ``False``
once a value for ``x`` is known. A value of ``None`` can not be used in that
way so it is rejected.

.. note:: We can use ``True`` in the ``Piecewise`` because ``True`` sympifies
          to ``S.true``. Sympifying ``None`` just gives ``None`` again which
          is not a valid symbolic SymPy object.

There are many other symbolic ``Boolean`` types in SymPy. The same
considerations about the differences between fuzzy bool and symbolic
``Boolean`` apply to all other sympy ``Boolean`` types. Just to give a

> different example there is ``Contains`` which represents the statement that an

Reviewer: ```suggestion
different example there is :class:`Contains`, which represents the statement that an
```

Author: Turns out `Contains` isn't documented anywhere

object is contained in a set:

    >>> from sympy import Reals, Contains
    >>> x = Symbol('x', real=True)
    >>> y = Symbol('y')
    >>> Contains(x, Reals)
    True
    >>> Contains(y, Reals)
    Contains(y, Reals)
    >>> Contains(y, Reals).subs(y, 1)
    True

The Python operator corresponding to ``Contains`` is ``in``. A quirk of ``in``
is that it can only evaluate to a ``bool`` (``True`` or ``False``) so if the
result is indeterminate then an exception will be raised:

    >>> from sympy import I
    >>> 2 in Reals
    True
    >>> I in Reals
    False
    >>> x in Reals
    True
    >>> y in Reals
    Traceback (most recent call last):
    ...
    TypeError: did not evaluate to a bool: (-oo < y) & (y < oo)

The exception can be avoided by using ``Contains(x, Reals)`` or
``Reals.contains(x)`` rather than ``x in Reals``.

Three-valued logic with fuzzy bools

> ===================================

Reviewer: Perhaps this section and the one below it could go in its own document about fuzzy booleans

Author: Yes, maybe it could. The problem is where should any of these documents go though?

The document I've written here feels like too much for the tutorial but all we have are the tutorial and the module docs. The module docs don't feel like a great place for a lengthy discussion either:

https://docs.sympy.org/latest/modules/assumptions/index.html
https://docs.sympy.org/latest/modules/core.html#module-sympy.core.assumptions

Author: Another thing that would be good to have somewhere would be a glossary page for terms like "vanilla symbol" and "fuzzy bool". It would also be good to have a documented that lists all `is_*` properties and explains what they mean. There isn't a clear place for those either.

Reviewer 2: We can create new rst files for long form guides like this. They can be linked to from the API docs. 

In general, I think we should try to organize our documentation using this system https://documentation.divio.com/.

Author: Yes that looks like a good structure. At the moment we only have two of the four quadrants.

:func:`~.simplify` is used:

    >>> simplify(sqrt(x**2))
    sqrt(x**2)

This is because SymPy will refuse to simplify this expression if the
simplification is not valid for *every* possible value of ``x``. By default
the symbol ``x`` is considered only to represent something roughly like an
arbitrary complex number and the obvious simplification here is only valid for
positive real numbers. Since ``x`` is not known to be positive or even real no
simplification of this expression is possible.

We can tell SymPy that a symbol represents a positive real number when
creating the symbol and then the simplification will happen automatically:

    >>> y = Symbol('y', positive=True)
    >>> sqrt(y**2)
    y

This is what is meant by "assumptions" in SymPy. If the symbol ``y``
is created with ``positive=True`` then SymPy will *assume* that it represents
a positive real number rather than an arbitrary complex or possibly infinite
number. That *assumption* can make it possible to simplify expressions or
might allow other manipulations to work. It is usually a good idea to be as
precise as possible about the assumptions on a symbol when creating it.

> The (old) assumptions system

Reviewer 2: This is part of a bigger discussion, but I don't think we should say "old" here. Our usage of "old" and "new" is sparse in the docs, enough that people are generally confused about which is which. But I think instead of trying to document it better, we should try to stop using those names, because

1. The "new" assumptions aren't even that new any more,
2. If we do end up merging them, then what would it be? "old-new"?
3. They are only descriptive for what they were historically.

Rather, we should have some better names. I'm not sure what the best ones are just yet. Something like "predicate assumptions" or "expression" assumptions for the new assumptions.

Author: I agree that we should use better names for this but if we invent new terms then we also need to have clear references linking those to the old terms.

I would go with "core assumptions" for the old assumptions.

============================

There are two sides to the assumptions system. The first side is that we can
declare assumptions on a symbol when creating the symbol. The other side is
that we can query the assumptions on any expression using the corresponding
``is_*`` attribute. For example:

    >>> x = Symbol('x', positive=True)
    >>> x.is_positive
    True

We can query assumptions on any expression not just a symbol:

    >>> expr = 1 + x**2
    >>> expr
    x**2 + 1
    >>> expr.is_positive
    True
    >>> expr.is_negative
    False

> The values given in an assumptions query use three-valued "fuzzy" logic. Any

Reviewer 2: It's a bit unfortunate that we call this "fuzzy" in functions like `fuzzy_and`, because fuzzy logic usually means something else https://en.wikipedia.org/wiki/Fuzzy_logic. Maybe a footnote should be here indicating that this is not to be confused with that.

real number but it also might *not* be. Without further information there is
no way to resolve the query ``x.is_positive``.

The other reason that an assumptions query might give ``None`` is just that
the assumptions system does not try very hard to answer complicated queries.
The system is intended to be fast and uses simple heuristic methods to
conclude a ``True`` or ``False`` answer in common cases. For example any sum
of positive terms is positive so:

    >>> from sympy import symbols
    >>> x, y = symbols('x, y', positive=True)
    >>> expr = x + y
    >>> expr
    x + y
    >>> expr.is_positive
    True

The last example is particularly simple so the assumptions system is able to
give a definite answer. If the sum involved a mix of positive or negative
terms it would be a harder query:

    >>> x = Symbol('x', real=True)
    >>> expr = 1 + (x - 2)**2
    >>> expr
    (x - 2)**2 + 1
    >>> expr.is_positive
    True
    >>> expr2 = expr.expand()
    >>> expr2
    x**2 - 4*x + 5
    >>> print(expr2.is_positive)
    None

Ideally that last example would give ``True`` rather than ``None`` because the
expression is always positive for any real value of ``x`` (and ``x`` has been

> assumed real). The assumptions system is intended to be efficient though: it

Reviewer 2: It's really three things: some algorithms aren't implemented because they would be too slow, some aren't implemented because they just aren't implemented yet, and some cases simply don't have algorithms at all (many assumptions are in general computationally undecidable).

Author: I'd make it four things:

1. Some queries are unknowable because there is not enough information provided e.g. `Symbol('x', real=True).is_positive`.

2. For some queries sufficient information is provided that a definite answer exists but the algorithmic problem of resolving the question for the class of queries is undecidable (Richardson's theorem).

3. For some queries algorithms exist but they are too slow and are deliberately not implemented.

4. In some cases simple algorithms could resolve the query but they haven't yet been implemented in sympy.

is expected many more complex queries will not be fully resolved. This is
because assumptions queries are primarily used internally by SymPy as part of
low-level calculations. Making the system more comprehensive would slow SymPy
down.

Note that in fuzzy logic giving an indeterminate result ``None`` is never a
contradiction. If it is possible to infer a definite ``True`` or ``False``
result when resolving a query then that is better than returning ``None``.
However a result of ``None`` is not a *bug*. Any code that uses the
assumptions system needs to be prepared to handle all three cases for any
query and should not presume that a definite answer will always be given.

> The assumptions system is not just for symbols or for complex expressions. It

Reviewer 2: I would remove this last paragraph and example. We already talk about using `S` earlier in the tutorial, and the details about `Basic` are implementation details that don't need to be in the tutorial.

e.g.:

    >>> from sympy import solve
    >>> solve('x**2 - 1')
    [-1, 1]

When creating symbols explicitly it would be possible to assign assumptions
that would affect the behaviour of :func:`~.solve`:

    >>> x = Symbol('x', positive=True)
    >>> solve(x**2 - 1)
    [1]

When using string input SymPy will create the expression and create all of the
symbolc implicitly so the question arises how can the assumptions be
specified? The answer is that rather than depending on implicit string
cponversion it is better to use the :func:`~.parse_expr` function explicitly
and then it is possible to provide assumptions for the symbols e.g.:

    >>> from sympy import parse_expr
    >>> parse_expr('x**2 - 1')
    x**2 - 1
    >>> eq = parse_expr('x**2 - 1', {'x':Symbol('x', positive=True)})
    >>> solve(eq)
    [1]

.. note:: The :func:`~.solve` function is unusual as a high level API in that it
          actually checks the assumptions on any input symbols (the unknowns)
          and uses that to tailor its output. The assumptions system otherwise
          affects low-level evaluation but is not necessarily handled
          explicitly by high-level APIs.

Predicates

> ==========

Reviewer 2: I would move this section in to the assumptions documentation rather than having it in the tutorial.

Reviewer 2: Same with most if not all the sections below this one. The tutorial should just be basic introductory stuff for new users.

     'extended_nonzero': True,
     'extended_positive': True,
     'extended_real': True,
     'finite': True,
     'hermitian': True,
     'imaginary': False,
     'infinite': False,
     'integer': True,
     'irrational': False,
     'negative': False,
     'noninteger': False,
     'nonnegative': True,
     'nonpositive': False,
     'nonzero': True,
     'positive': True,
     'rational': True,
     'real': True,
     'transcendental': False,
     'zero': False}

We can see that there are many more predicates listed than the two that were
used to create ``x``. This is because the assumptions system can infer some
predicates from combinations of other predicates. For example if a symbol is
declared with ``positive=True`` then it is possible to infer that it should
have ``negative=False`` because a positive number can never be negative.
Similarly if a symbol is created with ``integer=True`` then it is possible to
infer that is should have ``rational=True`` because every integer is a
rational number.

A full table of the possible predicates and their definitions is given below.

.. list-table:: Assumptions predicates for the (old) assumptions
    :widths: 20, 45, 35

>     :header-rows: 1

Reviewer 2: Can we generate the implication parts of this table automatically?
